packages/lib/src/components/UPI/components/UPIComponent/UPIComponent.tsx

```diff
@@ -0,0 +1,58 @@
+import { h, Fragment, RefObject } from 'preact';
+import { useMemo, useRef, useState } from 'preact/hooks';
+import VpaInput from '../VpaInput';
+import ContentSeparator from '../../../internal/ContentSeparator';
+import Button from '../../../internal/Button';
+import useCoreContext from '../../../../core/Context/useCoreContext';
+import getImage from '../../../../utils/get-image';
+import { PayButtonFunctionProps, UIElementStatus } from '../../../types';
+import { VpaInputHandlers } from '../VpaInput/VpaInput';
+
+interface UPIComponentProps {
+    payButton: (props: PayButtonFunctionProps) => h.JSX.Element;
+    ref: (ref: RefObject<typeof UPIComponent>) => void;
+    onChange: ({ data: VpaInputDataState, valid, errors, isValid: boolean }) => void;
+    onSubmit(): void;
+    onGenerateQrCodeClick(): void;
+}
+
+export default function UPIComponent({ onChange, onGenerateQrCodeClick, payButton }: UPIComponentProps): h.JSX.Element {
+    const { i18n, loadingContext } = useCoreContext();
+    const inputRef = useRef<VpaInputHandlers>(null);
+    const [status, setStatus] = useState<UIElementStatus>('ready');
+    const [isQrCodeFlow, setIsQrCodeFlow] = useState<boolean>(false);
+
+    this.setStatus = (status, isQrCodeFlow) => {
+        setIsQrCodeFlow(isQrCodeFlow);
+        setStatus(status);
+    };
+
+    this.showValidation = () => {
+        inputRef.current.validateInput();
+    };
+
+    const isSubmittingAndFlowIsQrCode = useMemo(() => status === 'loading' && isQrCodeFlow, [status, isQrCodeFlow]);
+
+    return (
+        <Fragment>
+            <VpaInput disabled={status === 'loading'} ref={inputRef} onChange={onChange} />
+
+            {payButton({
+                label: `${i18n.get('continue')}`,
+                status: isSubmittingAndFlowIsQrCode ? 'default' : status,
+                disabled: status === 'loading'
+            })}
+
+            <ContentSeparator label={i18n.get('qrCodeOrApp')} />
+
+            <Button
+                icon={getImage({ loadingContext: loadingContext, imageFolder: 'components/' })('qr_dark')}
+                status={isSubmittingAndFlowIsQrCode ? 'loading' : 'default'}
+                disabled={status === 'loading'}
```

Comment:
Can we set status to loading instead? Why create a new disabled state?
It's a bit confusing, since that is already the internal logic of the button, but we also use the status?

Reply:
This payment method differs a bit from the norm, as there are two buttons to submit the payment instead of one. The shopper can choose to 'Generate QR code', or to submit the VPA field .

So, when the shopper clicks to 'Generate QR code' , this button must show a loading state (as the user interacted with it), and the other button (the one that submits the VPA) must be disabled, and vice-versa. That is why the 'disabled' state is being used here. Otherwise both buttons would show 'loading' state, which isn't a good UX.

Reply:
I see, button already handles the logic of "loading", I had only read the logic of <PayButton>.
Maybe we should pass all this logic in the future to <Button> it would make it a cleaner interface.